### PR TITLE
Rescue inspect failures in Formatter (#2959)

### PR DESCRIPTION
**Pull request by xavier**
I took a quick shot at this. Are there any other failure case to account for?

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/formatter.ex [hunk 1]
```diff
@@ -186,10 +186,18 @@ defmodule ExUnit.Formatter do
...
-    |> inspect(pretty: true, width: width)
+    |> safe_inspect(pretty: true, width: width)
     |> String.replace("\n", "\n" <> @inspect_padding)
   end
 
+  defp safe_inspect(expr, opts \\ []) do
+    try do
+      inspect(expr, opts)
+    rescue
+      e -> "inspect failed: #{inspect e}"
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/formatter.ex]: If it fails, you can call it again but passing `structs: false` (it is the only reason it can fail).

**Comment by josevalim** [comment]
Thanks @xavier! This is a very good starting point and fixes part of the issue.

The bigger issue is that the Exception.format_banner and Exception.format_stacktrace must also be fixed! Try changing your snippet to this:

```elixir
  defmodule Failing do
    def __struct__ do
      %{key: 0}
    end
  
    defimpl Inspect do
      def inspect(struct, opts) when is_tuple(opts) do
        struct.unknown
      end
    end
  end
```

Because opts is never a tuple, everything should crash, including ExUnit itself, as the bad struct will also be in the stacktrace and error message.

Actually, I am starting to think we should rather change inspect behaviour to log, instead of raising an error around here:

https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/inspect/algebra.ex#L203

This way we wouldn't need to patch a bunch of functions to make them safe... Thoughts?

/cc @ericmj and @alco?

**Comment by xavier** [comment]
Thanks for the feedback.

If this ExUnit enhancement also involves changes to the stdlib then I suppose that this problem should be better solved at the source by not raising an ArgumentError in the first place, as you proposed.

**Comment by josevalim** [comment]
Alright! So if you would like, let's modify this pull request so we use `IO.puts :stderr, msg` inside Inspect's to_doc. :) But I think we should keep the test case in ExUnit, to guarantee it will work as we expect!

**Comment by ericmj** [comment]
We should definitely not log or `IO.puts` from core parts such as `Inspect`, that feels very wrong. It is much better to try to handle this explicitly in ExUnit, possibly with an option to `Inspect`.

**Comment by xavier** [comment]
How about having inspect return the error itself as a string, possibly tagged, e.g. return a string such as `"{:inspect_failed, reason}"`

This is of course semantically debatable but I could probably live with this if inspect failures are considered corner cases. I'm actually not sure how common these errors are, I don't think I ever encountered one but I suspect it may happen when dealing with custom inspect algebra?

**Comment by josevalim** [comment]
@ericmj there is no way we can fix this issue without changing Elixir core. I think that's the biggest issue here, inspect is unsafe, and I think it should be made safe.

@xavier your suggestions of wrapping it in an #Inspect.Error<...> that would still be returned as a string sounds like a good one. @ericmj, what do you think?

**Comment by ericmj** [comment]
Im fine with changing core as long as we don't print from functions that really shouldn't print anything. We shouldn't make `Inspect` safe either, it will definitely hide bugs if inspect errors are returned as strings instead of raising.

Don't know why we just can't rescue the error in ExUnit. If we want the `structs: false` output we can add that as a field to the exception.

**Comment by josevalim** [comment]
Ok. So I just added `safe: true` to inspect itself. This will make it easier and consistent to add safe_inspect to both Exception and the ExUnit.Formatter: https://github.com/elixir-lang/elixir/commit/23d318fa0c8fb94df523897b98e696b11123a253

@xavier I think we can go ahead with your patch then, we just need to use the new safe option and guarantee the functions in Exception are also safe with that new test case.

Thanks everyone!

**Comment by josevalim** [comment]
Hello @xavier! We have decided to make safe: true the default. I have merged your test cases and thank you for your explorations on this issue, it definitely helped us find a solution faster!

**Comment by xavier** [comment]
:+1:

**Pull request closed without merging**